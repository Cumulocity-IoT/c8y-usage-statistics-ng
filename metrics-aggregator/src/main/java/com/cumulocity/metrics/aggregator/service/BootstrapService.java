package com.cumulocity.metrics.aggregator.service;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Service;
import com.cumulocity.microservice.subscription.model.MicroserviceSubscriptionAddedEvent;






/**
 * This is a service to fetch subtenants etc
 * 
 * @author Marco Stoffel
 *
 */

@Service
public class BootstrapService {

    @Autowired
    DeviceMetricsAggregationService deviceMetricsAggregationService;
 
    @Autowired
    StartupUtils utils;

    private static final Logger log = LoggerFactory.getLogger(BootstrapService.class);

    @EventListener
	public void initialize(MicroserviceSubscriptionAddedEvent event) {
		log.info("Tenant Sub: " + event.toString());
        utils.setCurrentTenant(event.getCredentials().getTenant());
        utils.getTenants();
        deviceMetricsAggregationService.createDailyDeviceStatistics();
	}

}
