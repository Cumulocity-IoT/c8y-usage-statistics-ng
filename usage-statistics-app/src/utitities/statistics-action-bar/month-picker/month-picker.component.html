
    <form class="container" style="width: 180px;">
        <div class="form-group datepicker" style="margin:0">
            <input 
        [(ngModel)]="selectedDate" 
        autocomplete="off" 
        class="form-control" 
        name="date" 
        bsDatepicker 
        [bsConfig]="{dateInputFormat: 'MMMM/YYYY', adaptivePosition: true }" 
        (onShown)="onOpenCalendar($event)"
        (onHidden)="update()"
        [maxDate]="maxDate"
        style="cursor: pointer; color: black;"
        readonly
        >
        </div>
     
    </form>
