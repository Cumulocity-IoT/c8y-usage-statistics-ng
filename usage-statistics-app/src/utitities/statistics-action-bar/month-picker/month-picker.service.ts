import { Injectable } from "@angular/core";
import { Subject } from "rxjs";


@Injectable({
    providedIn: 'root'
})
export class MonthPickerService {
    selectedDate: Date;    
    daysInMonth: number;
    dateChanged = new Subject();   
}