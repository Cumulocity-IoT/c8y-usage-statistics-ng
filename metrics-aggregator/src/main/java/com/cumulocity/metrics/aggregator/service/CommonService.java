package com.cumulocity.metrics.aggregator.service;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Service;

import com.cumulocity.metrics.aggregator.model.microservice.Tenants;
import com.cumulocity.metrics.aggregator.model.microservice.Tenants.Tenant;
import com.cumulocity.microservice.subscription.model.MicroserviceSubscriptionAddedEvent;
import com.cumulocity.microservice.subscription.service.MicroserviceSubscriptionsService;
import com.cumulocity.rest.representation.CumulocityMediaType;
import com.cumulocity.sdk.client.RestConnector;






/**
 * This is a service to fetch subtenants etc
 * 
 * @author Marco Stoffel
 *
 */
@Service
public class CommonService {

    @Autowired
    RestConnector restConnector;
    

    @Autowired
	MicroserviceSubscriptionsService subscriptionsService;

    private static final Logger log = LoggerFactory.getLogger(CommonService.class);

    private String currentTenant = ""; 
    private List<String> tenantList;

    public String getCurrentTenant() {
        return currentTenant;
    }

    public List<String> getTenantList() {
        return tenantList;
    }

    @EventListener
	public void initialize(MicroserviceSubscriptionAddedEvent event) {
		log.info("Tenant Sub: " + event.toString());
        this.currentTenant = event.getCredentials().getTenant();
		log.info("CurrentTenant: " + this.currentTenant);

        this.tenantList = subscriptionsService.callForTenant(this.currentTenant, this::getSubTenants).get();
        this.tenantList.add(currentTenant);
        log.info(tenantList.toString());
	}

    private Optional<List<String>> getSubTenants(){
        Tenants tenants = restConnector.get(
                "/tenant/tenants?pageSize=2000&withTotalPages=true&withApps=false",
                CumulocityMediaType.APPLICATION_JSON_TYPE,Tenants.class);
            ArrayList<String> tenantIds = new ArrayList<String>();
            for (Tenant tenant : tenants.getTenants()){
                tenantIds.add(tenant.getId());
            }
            return Optional.of(tenantIds);
    }

}
