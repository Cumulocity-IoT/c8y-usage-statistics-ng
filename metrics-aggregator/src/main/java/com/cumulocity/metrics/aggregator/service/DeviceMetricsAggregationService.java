package com.cumulocity.metrics.aggregator.service;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.time.Instant;
import java.time.Period;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.context.annotation.Configuration;
import org.springframework.scheduling.annotation.EnableScheduling;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import com.cumulocity.sdk.client.RestConnector;
import com.cumulocity.sdk.client.SDKException;


import com.cumulocity.metrics.aggregator.model.device.DeviceClassConfiguration;
import com.cumulocity.metrics.aggregator.model.device.DeviceStatistics;
import com.cumulocity.metrics.aggregator.model.device.DeviceStatistics.Statistic;
import com.cumulocity.metrics.aggregator.model.device.DeviceStatisticsAggregation;
import com.cumulocity.metrics.aggregator.model.device.DeviceClassConfiguration.DeviceClass;
import com.cumulocity.microservice.subscription.service.MicroserviceSubscriptionsService;
import com.cumulocity.model.option.OptionPK;
import com.cumulocity.rest.representation.CumulocityMediaType;
import com.cumulocity.rest.representation.tenant.OptionRepresentation;
import com.cumulocity.sdk.client.option.TenantOptionApi;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;



/**
 * This is a service to aggregate device an microservices metrics to display in
 * the usage statistics app
 * It will iterate over all subscribed tenants and aggregate the metrics to an
 * overview
 * 
 * @author Marco Stoffel
 *
 */
@Configuration
@EnableScheduling
@Service
public class DeviceMetricsAggregationService {

	private static final Logger log = LoggerFactory.getLogger(DeviceMetricsAggregationService.class);

	private static final String OPTION_CATEGORY_CONFIGURATION = "configuration";
	private static final String OPTION_KEY = "device.statistics.class.details";

	private Map<String, DeviceClassConfiguration> allDeviceClassConfiguration = new HashMap<String, DeviceClassConfiguration>();
	//private Map<Date,DeviceStatisticsAggregation> deviceStatisticsAggregationCurrentMonth = new HashMap<Date,DeviceStatisticsAggregation>();


	DeviceStatisticsAggregation dailyDeviceStatisticsAggregation = new DeviceStatisticsAggregation();
	Instant dailyDeviceStatisticsAggregationLastRun=  Instant.now().minus(Period.ofDays(2));
	

	@Autowired
	RestConnector restConnector;

	@Autowired
	MicroserviceSubscriptionsService subscriptionsService;


	@Autowired
	TenantOptionApi tenantOptionApi;

	@Autowired
	ObjectMapper objectMapper;


	private List<String> tenantList;

	public List<String> getTenantList() {
		return tenantList;
	}

	public void setTenantList(List<String> tenantList) {
		this.tenantList = tenantList;
	}

	private DateFormat df = new SimpleDateFormat("yyyy-MM-dd");

	public Map<String, DeviceStatistics> getDeviceStatisticsOverview(String type, Date statDate) {
		HashMap<String, DeviceStatistics> dsMap = new HashMap<String, DeviceStatistics>();
			for (String currentTenant : this.getTenantList()) {
				log.info("Get Statistics for Tenant: " + currentTenant);
				
				DeviceStatistics dspage = restConnector.get(
						"/tenant/statistics/device/" + currentTenant + "/" + type + "/" + df.format(statDate)
								+ "?pageSize=2000&withTotalPages=true",
						CumulocityMediaType.APPLICATION_JSON_TYPE, DeviceStatistics.class);
				dspage.setDaysInMonth(statDate,type);
				log.debug("Statistics: " + dspage.getStatistics().toString());
				dsMap.put(currentTenant, dspage);
			}
		return dsMap;
	}

	public Map<String, DeviceClassConfiguration> getAllDeviceClassConfiguration(boolean useTenantDeviceClasses) {
		log.info("# Fetching all deviceClassConfigurations, useTenantDeviceClasses: " + useTenantDeviceClasses);
		this.allDeviceClassConfiguration = new HashMap<String, DeviceClassConfiguration>();
		for (String currentTenant : this.getTenantList()) {
				if (useTenantDeviceClasses) {
					this.allDeviceClassConfiguration.put(currentTenant, getDeviceClasseConfiguration(currentTenant));
				} else {
					this.allDeviceClassConfiguration.put(currentTenant, new DeviceClassConfiguration());
				}

		}
		return allDeviceClassConfiguration;
	}

	private DeviceClassConfiguration getDeviceClasseConfiguration(String currentTenant) {
		log.info("Get DeviceClassConfiguration for Tenant: " + currentTenant);
		final OptionPK option = new OptionPK();
		option.setCategory(OPTION_CATEGORY_CONFIGURATION);
		option.setKey(OPTION_KEY);
		OptionRepresentation optionRepresentation = new OptionRepresentation();
		List<DeviceClassConfiguration.DeviceClass> dcl = new ArrayList<DeviceClassConfiguration.DeviceClass>();
		DeviceClassConfiguration deviceClassConfiguration = new DeviceClassConfiguration();
		try {
			optionRepresentation = tenantOptionApi.getOption(option);
			dcl = new ObjectMapper().readValue(
					optionRepresentation.getValue(),
					new TypeReference<List<DeviceClassConfiguration.DeviceClass>>() {
					});
			deviceClassConfiguration.setDeviceClasses(dcl);
		} catch (SDKException | JsonProcessingException e) {
			log.error("Could not get DeviceClassConfiguration form tenant options using default config.");
		}
		return deviceClassConfiguration;
	}

	@Cacheable(value = "deviceCache", key = "#statDate.toString() + '-' + #type + '-' + #includeSubtenants", condition = "#useTenantDeviceClasses == false")
	public DeviceStatisticsAggregation getAggregatedDeviceClassStatistics(String type, Date statDate,
			boolean includeSubtenants, boolean useTenantDeviceClasses) {
		// Get all Device Statistics for all Tenants

		Map<String, DeviceStatistics> deviceStatisticsMap = this.getDeviceStatisticsOverview(type, statDate);
		Map<String, DeviceClassConfiguration> deviceClassConfigurationMap = this
				.getAllDeviceClassConfiguration(useTenantDeviceClasses);
		// Get all device class configurations
		return getAggregatedDevicesPerClass(deviceStatisticsMap, deviceClassConfigurationMap, includeSubtenants, type);
	}



	private DeviceStatisticsAggregation getAggregatedDevicesPerClass(
			Map<String, DeviceStatistics> deviceStatisticsMap,
			Map<String, DeviceClassConfiguration> deviceClassesMap, boolean includeSubtenants, String type) {

		final Boolean incSubtenants = includeSubtenants;
		// Return object to gather statistics
		DeviceStatisticsAggregation deviceStatisticsAggregation = new DeviceStatisticsAggregation();

		// Loop through all tenans deviceStatistics
		Iterator<Entry<String, DeviceStatistics>> dsIterator = deviceStatisticsMap.entrySet().iterator();
		while (dsIterator.hasNext()) {
			Map.Entry<String, DeviceStatistics> pair = (Entry<String, DeviceStatistics>) dsIterator
					.next();
			String tenant = pair.getKey();
			DeviceStatistics ds = pair.getValue();
			DeviceClassConfiguration deviceClassConfigCurrentTenant = deviceClassesMap.get(tenant);
			// Write amount of devices to total
			deviceStatisticsAggregation.addTotalDevicesCount(ds.getStatistics().size());
			// Loop through all device statistics of current tenant
			DeviceStatisticsAggregation.TenantAggregation ta = new DeviceStatisticsAggregation.TenantAggregation();
			if (includeSubtenants) {
				deviceStatisticsAggregation.getTenantAggregation().put(tenant, ta);
			}
			ds.getStatistics().forEach(statOfSingleDevice -> {
				deviceStatisticsAggregation.addTotalMeas(statOfSingleDevice.getCount());
				if (incSubtenants) {
					ta.addToMeas(statOfSingleDevice.getCount());
					ta.setDeviceClasses(deviceClassConfigCurrentTenant);
					ta.setDevicesCount(ds.getStatistics().size());
					updateDeviceClass(ta.getDeviceClasses(), statOfSingleDevice.getCount(), ds.getDaysInMonth());
				}
				updateDeviceClass(deviceStatisticsAggregation.getTotalDeviceClasses(),
						statOfSingleDevice.getCount(), ds.getDaysInMonth());
			});

		}
		return deviceStatisticsAggregation;
	}

	
	private void updateDeviceClass(DeviceClassConfiguration deviceClassConfiguration, int count, int daysInMonth) {
		float avgMea = count / daysInMonth;
		Iterator<DeviceClass> idc = deviceClassConfiguration.getDeviceClasses().iterator();
		while (idc.hasNext()) {
			DeviceClass dc = idc.next();
			if ((dc.getAvgMaxMea() instanceof String && dc.getAvgMaxMea().equals("INFINITY")
					&& avgMea >= Float.valueOf(dc.getAvgMinMea()))
					||
					(dc.getAvgMaxMea() instanceof Integer && avgMea >= Float.valueOf(dc.getAvgMinMea())
							&& avgMea < Float.valueOf((int) dc.getAvgMaxMea()))) {
				dc.incrementCount();
			}
			;
		}
	}


	public DeviceStatisticsAggregation getDailyStatistics(boolean omitCache){
		Instant now = Instant.now();
		long hours = ChronoUnit.HOURS.between(this.dailyDeviceStatisticsAggregationLastRun, now );
		// fetch only when older than 12 hours
		if (omitCache || hours > 24 )	 {
			log.info("Getting Daily stats.");
			createDailyDeviceStatistics();
		}
		return this.dailyDeviceStatisticsAggregation;
	}


	
	public void createDailyDeviceStatistics(){

			this.dailyDeviceStatisticsAggregation = new DeviceStatisticsAggregation();
			Map<String,Integer> devicesDailyAggregation = new HashMap<String,Integer>();
			Calendar cal = Calendar.getInstance();
			
			int dayOfMonth = cal.get(Calendar.DAY_OF_MONTH);
			cal.set(Calendar.HOUR_OF_DAY, 0);
			cal.set(Calendar.MINUTE, 0);
			cal.set(Calendar.SECOND, 0);
			// cal.set(Calendar.MONTH, 1);
			// dayOfMonth = 28;
			
			// get all device meas from start of month
			for (int i = 1; i <= dayOfMonth; i++){
				cal.set(Calendar.DAY_OF_MONTH, i);
				Date statDate = cal.getTime();
				
				log.info("Get daily statistic: "+ statDate.toString());
				//this.deviceStatisticsAggregationCurrentMonth.put(statDate, getAggregatedDeviceClassStatistics("daily", statDate ,  false, false));
				Map<String, DeviceStatistics> devstat = this.getDeviceStatisticsOverview("daily", statDate);
	
				for (var entry : devstat.entrySet()) {
					List<Statistic> ls =entry.getValue().getStatistics();
					for (Statistic st : ls) {
						devicesDailyAggregation.put(
							st.getDeviceId(), 
							devicesDailyAggregation.getOrDefault(st.getDeviceId(), dayOfMonth) +st.getCount()
							);
					}
				}
			}
			
			// Get device classes
			this.dailyDeviceStatisticsAggregation.setTotalDeviceCount(devicesDailyAggregation.size());
			DeviceClassConfiguration dailyDeviceClasses = new DeviceClassConfiguration();
			for (var device : devicesDailyAggregation.entrySet()){
				this.dailyDeviceStatisticsAggregation.addTotalMeas(device.getValue());
				this.updateDeviceClass(dailyDeviceClasses, device.getValue(), dayOfMonth);
			}
			this.dailyDeviceStatisticsAggregation.setTotalDeviceClasses(dailyDeviceClasses);
			this.dailyDeviceStatisticsAggregationLastRun = Instant.now();
			//log.info("Daily Stat: " + this.deviceStatisticsAggregationCurrentMonth.toString());

    }


	@Scheduled(cron = "0 15 0 * * ?")
	public void scheduleDailyStatistics(){
		subscriptionsService.runForEachTenant(()->{
			this.createDailyDeviceStatistics();
		});
		log.info("ScheduledDailyStatistics");
	}
}
