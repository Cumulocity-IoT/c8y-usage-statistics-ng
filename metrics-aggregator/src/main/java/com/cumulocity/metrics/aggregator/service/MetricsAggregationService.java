package com.cumulocity.metrics.aggregator.service;

import java.math.BigInteger;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.time.YearMonth;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.context.event.ApplicationReadyEvent;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Service;

import com.cumulocity.sdk.client.RestConnector;
import com.cumulocity.sdk.client.SDKException;
import com.cumulocity.metrics.aggregator.model.DeviceClassConfiguration;
import com.cumulocity.metrics.aggregator.model.DeviceStatistics;
import com.cumulocity.metrics.aggregator.model.DeviceStatisticsAggregation;
import com.cumulocity.metrics.aggregator.model.DeviceClassConfiguration.DeviceClass;
import com.cumulocity.microservice.subscription.model.MicroserviceSubscriptionAddedEvent;
import com.cumulocity.microservice.subscription.service.MicroserviceSubscriptionsService;
import com.cumulocity.model.option.OptionPK;
import com.cumulocity.rest.representation.CumulocityMediaType;
import com.cumulocity.rest.representation.tenant.OptionRepresentation;
import com.cumulocity.rest.representation.tenant.TenantRepresentation;
import com.cumulocity.sdk.client.option.TenantOptionApi;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import javassist.expr.Instanceof;

import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

/**
 * This is a service to aggregate device an microservices metrics to display in
 * the usage statistics app
 * It will iterate over all subscribed tenants and aggregate the metrics to an
 * overview
 * 
 * @author Marco Stoffel
 *
 */
@Service
public class MetricsAggregationService {

	private static final Logger log = LoggerFactory.getLogger(MetricsAggregationService.class);

	private static final String OPTION_CATEGORY_CONFIGURATION = "configuration";
	private static final String OPTION_KEY = "device.statistics.class.details";

	private Map<String, DeviceClassConfiguration> allDeviceClassConfiguration = new HashMap<String, DeviceClassConfiguration>();

	private int subscribedTenantCount = 0;
	private boolean includeSubtenants = false;

	@Autowired
	RestConnector restConnector;

	@Autowired
	MicroserviceSubscriptionsService subscriptionsService;

	@Autowired
	TenantOptionApi tenantOptionApi;

	@Autowired
	ObjectMapper objectMapper;

	@EventListener
	public void initialize(MicroserviceSubscriptionAddedEvent event) {
		// Optional<TenantRepresentation> deployedTenant = getCurrentTenant();
		// System.out.println("Current Tenant: " + deployedTenant.get().getDomain());
		String currentTenant = event.getCredentials().getTenant();
		subscriptionsService.runForTenant(currentTenant, () -> {
			this.subscribedTenantCount++;
			this.allDeviceClassConfiguration.put(currentTenant, getDeviceClasseConfiguration(currentTenant));
			log.info("Tenant: " + currentTenant + " subscribed. subscribedTenantCount: " + subscribedTenantCount
					+ " Fetching DeviceClassConfiguration.");
		});
	}

	// @EventListener(ApplicationReadyEvent.class)
	// public void doSomethingAfterStartup() {
	// 	Optional<TenantRepresentation> currentTenant= getCurrentTenant();
	// 	System.out.println("Current Tenant: " +currentTenant.get());
	// }

	private DateFormat df = new SimpleDateFormat("yyyy-MM-dd");

	public Map<String, DeviceStatistics> getDeviceStatisticsOverview(String type, Date statDate) {

		HashMap<String, DeviceStatistics> dsMap = new HashMap<String, DeviceStatistics>();
		subscriptionsService.runForEachTenant(() -> {
			String currentTenant = subscriptionsService.getTenant();
			log.info("Get Statistics for Tenant: " + currentTenant);
			Response rs = restConnector.get(
					"/tenant/statistics/device/" + currentTenant + "/" + type + "/" + df.format(statDate)
							+ "?pageSize=2&withTotalPages=true",
					MediaType.APPLICATION_JSON_TYPE);
			DeviceStatistics dspage = restConnector.get(
					"/tenant/statistics/device/" + currentTenant + "/" + type + "/" + df.format(statDate)
							+ "?pageSize=2000&withTotalPages=true",
					MediaType.APPLICATION_JSON_TYPE, DeviceStatistics.class);
			dspage.setDaysInMonth(statDate);
			log.debug("Statistics: " + dspage.getStatistics().toString());
			dsMap.put(currentTenant, dspage);
		});
		return dsMap;
	}

	public Map<String, DeviceClassConfiguration> getAllDeviceClassConfiguration(boolean omitCache) {
		if (omitCache) {
			log.info("# Fetching deviceClassRepresentation, omitCache: " + omitCache);
			this.allDeviceClassConfiguration = new HashMap<String, DeviceClassConfiguration>();
			subscriptionsService.runForEachTenant(() -> {
				String currentTenant = subscriptionsService.getTenant();
				this.allDeviceClassConfiguration.put(currentTenant, getDeviceClasseConfiguration(currentTenant));
			});
		}
		return allDeviceClassConfiguration;
	}

	private DeviceClassConfiguration getDeviceClasseConfiguration(String currentTenant) {
		log.info("Get DeviceClassDefinition for Tenant: " + currentTenant);
		final OptionPK option = new OptionPK();
		option.setCategory(OPTION_CATEGORY_CONFIGURATION);
		option.setKey(OPTION_KEY);
		OptionRepresentation optionRepresentation = new OptionRepresentation();
		List<DeviceClassConfiguration.DeviceClass> dcl = new ArrayList<DeviceClassConfiguration.DeviceClass>();
		DeviceClassConfiguration dcd = new DeviceClassConfiguration();
		try {
			optionRepresentation = tenantOptionApi.getOption(option);
			dcl = new ObjectMapper().readValue(
					optionRepresentation.getValue(),
					new TypeReference<List<DeviceClassConfiguration.DeviceClass>>() {
					});
			dcd.setDeviceClasses(dcl);
		} catch (SDKException | JsonProcessingException e) {
			log.error("Could not get DeviceClassConfiguration form tenant options using default config.");
		}
		return dcd;
	}

	public DeviceStatisticsAggregation getAggregatedDeviceClassStatistics(String type, Date statDate, boolean includeSubtenants) {
		Map<String, DeviceStatistics> deviceStatisticsMap = this.getDeviceStatisticsOverview(type, statDate);
		boolean omitCache = false;
		this.includeSubtenants = includeSubtenants;
		if (deviceStatisticsMap.size() != this.allDeviceClassConfiguration.size()) {
			log.info("################## Size DeviceClasses and DeviceStatistics uneven. Fetching DeviceClassesConfig again.");
			omitCache = true;
		}
		Map<String, DeviceClassConfiguration> deviceClassesMap = this.getAllDeviceClassConfiguration(omitCache);
		return getAggregatedDevicesPerClass(deviceStatisticsMap, deviceClassesMap);
	}

	private DeviceStatisticsAggregation getAggregatedDevicesPerClass(
			Map<String, DeviceStatistics> deviceStatisticsMap,
			Map<String, DeviceClassConfiguration> deviceClassesMap) {

		// Return object to gather statistics
		DeviceStatisticsAggregation deviceStatisticsAggregation = new DeviceStatisticsAggregation();
		
		// Loop through all tenans deviceStatistics
		Iterator<Entry<String, DeviceStatistics>> dsIterator = deviceStatisticsMap.entrySet().iterator();
		while (dsIterator.hasNext()) {
			Map.Entry<String, DeviceStatistics> pair = (Entry<String, DeviceStatistics>) dsIterator
					.next();
			String tenant= pair.getKey();
			DeviceStatistics ds = pair.getValue();
			DeviceClassConfiguration dcc = deviceClassesMap.get(tenant);
			deviceStatisticsAggregation.addTotalDevicesCount(ds.getStatistics().size());
			ds.getStatistics().forEach(devs ->{
				deviceStatisticsAggregation.addTotaMeas(devs.getCount());
				if (this.includeSubtenants) {
						DeviceStatisticsAggregation.TenantAggregation ta = new DeviceStatisticsAggregation.TenantAggregation();
						deviceStatisticsAggregation.getTenantAggregation().put(tenant, ta);
						ta.addToMeas(devs.getCount());
						ta.setDeviceClasses(dcc.getDeviceClasses());
						ta.setDevicesCount(ds.getStatistics().size());
						updateDeviceClass(dcc, devs.getCount(), ds.getDaysInMonth());
					}
					updateDeviceClass(deviceStatisticsAggregation.getTotalDeviceClasses(), 
						devs.getCount(), ds.getDaysInMonth());
			});	

		}
		return deviceStatisticsAggregation;
	}

	private void updateDeviceClass(DeviceClassConfiguration dcc, int count, int daysInMonth){
		float avgMea = count / daysInMonth;
		Iterator<DeviceClass> idc = dcc.getDeviceClasses().iterator();
		while (idc.hasNext()) {
			DeviceClass dc = idc.next();
			if (
				(dc.getAvgMaxMea() instanceof String && dc.getAvgMaxMea().equals("INFINITY") && avgMea >= Float.valueOf(dc.getAvgMinMea()))
				|| 
				(dc.getAvgMaxMea() instanceof Integer && avgMea >= Float.valueOf(dc.getAvgMinMea()) && avgMea < Float.valueOf((int)dc.getAvgMaxMea()))
			){
				dc.incrementCount();
			};
		}
	}

	public Optional<TenantRepresentation> getCurrentTenant() {
		try {
			return Optional.ofNullable(restConnector.get("/tenant/currentTenant", CumulocityMediaType.APPLICATION_JSON_TYPE, TenantRepresentation.class));
		} catch (final SDKException e) {
			log.error("Tenant#getCurrentTenant operation resulted in " + e.getMessage(), e);
		}
		return Optional.empty();
	}

	
}
