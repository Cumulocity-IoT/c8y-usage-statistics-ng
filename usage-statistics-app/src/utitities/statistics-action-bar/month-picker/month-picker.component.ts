import { Component } from '@angular/core';
import { DATE_FORMAT_MONTH } from '../../../common.service';
import { MonthPickerService } from './month-picker.service';
const moment = require('moment');

@Component({
  selector: 'month-picker',
  templateUrl: './month-picker.component.html',
  styleUrls: ['./month-picker.component.css']
})
export class MonthPickerComponent {
  selectedDate: string;
  maxDate: Date;

  constructor(
    private monthPickerService: MonthPickerService
  ) {
    this.selectedDate = this.monthPickerService.selectedDate ?
      moment(this.monthPickerService.selectedDate).format(DATE_FORMAT_MONTH) :
      moment().subtract(1, 'months').format(DATE_FORMAT_MONTH);
    this.update()
  }

  ngOnInit() {
    const date = new Date();
    this.maxDate = new Date(date.getFullYear(), date.getMonth() - 1, 1);
  }
  onOpenCalendar(container) {
    container.monthSelectHandler = (event: any): void => container._store.dispatch(container._actions.select(event.date));
    container.setViewMode('month');
  }

  update() {
    this.monthPickerService.selectedDate = this.monthPickerService.selectedDate ? new Date(this.selectedDate) : this.getLastMonth();
    this.monthPickerService.daysInMonth = moment(this.selectedDate, DATE_FORMAT_MONTH).daysInMonth() 
    this.monthPickerService.dateChanged.next(this.monthPickerService.selectedDate)
  }

  private getLastMonth() {
    const today = new Date();
    const lastMonthFirstDay = new Date(today.getFullYear(), today.getMonth() - 1, 1);
    return lastMonthFirstDay;
  }
}
