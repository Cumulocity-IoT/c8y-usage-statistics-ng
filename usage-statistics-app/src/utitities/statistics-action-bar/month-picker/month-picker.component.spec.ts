import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { CoreModule } from '@c8y/ngx-components';
import { Subject } from 'rxjs';
import { MonthPickerComponent } from './month-picker.component';
import { MonthPickerService } from './month-picker.service';

describe('Test MonthPickerComponent', () => {
  let component: MonthPickerComponent;
  let fixture: ComponentFixture<MonthPickerComponent>;
  let monthPickerService: MonthPickerService;

  const monthPickerServiceMock = {
    selectedDate: new Date("2023-01-01"),
    daysInMonth: 31,
    dateChanged: new Subject()
  }

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule, CoreModule.forRoot()],
      declarations: [MonthPickerComponent],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [
        { provide: MonthPickerService, useValue: monthPickerServiceMock },
    ]
    });
    fixture = TestBed.createComponent(MonthPickerComponent);
    component = fixture.componentInstance;
    monthPickerService = TestBed.inject(MonthPickerService);
  });

  test('should be defined', () => {
    expect(fixture).toBeDefined();
  });

  test('should load expected component', () => {
    expect(fixture).toMatchSnapshot();
  });

  test('should define selectedDate and maxDate' , () => {
    fixture.detectChanges();
    expect(component.selectedDate).toBeDefined();
    expect(component.maxDate).toBeDefined();
  })

});