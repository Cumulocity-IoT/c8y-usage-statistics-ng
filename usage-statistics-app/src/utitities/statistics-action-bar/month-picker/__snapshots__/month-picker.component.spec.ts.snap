// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MonthPickerComponent should load expected component 1`] = `
<month-picker
  monthPickerService={[Function Object]}
  selectedDate={[Function String]}
>
  <form
    class="container"
    novalidate=""
    style="width: 180px;"
  >
    <div
      class="form-group datepicker"
      style="margin: 0;"
    >
      <input
        autocomplete="off"
        bsdatepicker=""
        class="form-control"
        name="date"
        readonly=""
        style="cursor: pointer; color: black;"
      />
    </div>
  </form>
</month-picker>
`;
